### ISSUE TITLE
OpenSSL 1.1.1 compile errors when building Ruby

### ISSUE THREAD

**Issue by MSP-Greg** [msys2/MINGW-packages#4467]
As you may recall, I've got a 1.1.1 package, and successfully build Ruby and two Ruby gems that also compile with OpenSSL.

When using the new package, I've got many compile issues/errors, please see [this Appveyor log](https://ci.appveyor.com/project/MSP-Greg/ruby-loco/build/1207#L2350).  Currently supported Ruby versions are being compiled with MSYS2/MinGW 1.0.2, and custom 1.1.0 and 1.1.1 packages, so I doubt the issue is on the Ruby side.

My 1.1.1 package was based on a previous 1.1.0 package that I and others have used building Ruby.  Hence, it was quite a while ago when I made the determination what patches to carry forward from 1.0.2 to 1.1.0.  IOW, I can't remember.  Also, OpenSSL does build on VC, and also cross-compiles on MinGW, so I knew that there could be updates that removed the need for patches.

I'm not a strong c type, so I'm not sure what's causing these issues, but I suspect it is one of the old patches from 1.0.2.

I'll always be building OpenSSL when it hits beta, but it would be nice to use a stock package for normal CI...

Thanks, Greg

**Comment by MSP-Greg** [comment]
@JPeterMugaas @lazka

Any ideas?  I did build the package locally (64 bit only), and all tests passed.  On Appveyor, I just installed the package for the Ruby build, it shows up [here](https://ci.appveyor.com/project/MSP-Greg/ruby-loco/build/1207?fullLog=true#L92) using a `pacman -Qs` command...

**Comment by lazka** [comment]
The build log error looks similar to other errors where "#include <windows.h>" is now missing before the openssl include, which is a new requirement with openssl1.1.

> When using the new package, I've got many compile issues/errors,

Which new package?

**Comment by MSP-Greg** [comment]
@lazka 

> The build log error looks similar to

Apologies.  The link in the first message was incorrect (it was the current Appveyor build job when written, it is now not).  Updated.  Please look again at the errors.

> Which new package?

[mingw-w64-x86_64-openssl 1.1.1-3](https://ci.appveyor.com/project/MSP-Greg/ruby-loco/build/1207?fullLog=true#L92)

The build log is a log from a rep that builds Ruby master/trunk (CRON job, three times a day, just like the Travis rvm Ruby head jobs), tests it, and saves it for other CI.

**Comment by lazka** [comment]
Could be due to https://github.com/Alexpux/MINGW-packages/pull/4441

The error is still the same, missing windows.h include before including openssl/engine.h

**Comment by MSP-Greg** [comment]
@lazka

> Could be due to #4441...

It could, but it's not.  Used the same code in 1.1.0, which I've been using since it's release.  I think I changed to 1.1.1 around pre8, and I built OpenSSL several times from master until release.

Some of the problems with 1.1.1-3 may be due to the code/packages that it is being built with it.  Ruby needed some changes when the transition from 1.0.2 to 1.1.0 happened.  As I recall, the only change needed was due to the name changes in the dll's/libraries.

There may also be issues with hostnames (test failures, not build issues).  1.1.1 changed some code for them, and I found some issues that they fixed, as the changes were incompatible with the 1.1.0 api.

Anyway, I've got a system that builds successfully with 1.0.2, 1.1.0, and the 1.1.1 package I use.  It does not work with 1.1.1-3.

**Comment by lazka** [comment]
OK. Is there a link somewhere to download your 1.1.1 build?

**Comment by MSP-Greg** [comment]
@lazka

Yes, in the other issue... (Maybe I shouldn't have opened this one, but the other seemed to be somewhat different)

https://github.com/Alexpux/MINGW-packages/issues/4458#issuecomment-424937023

All the artifacts from the build are [here](https://ci.appveyor.com/project/MSP-Greg/ruby-makepkg-mingw/build/7/artifacts), including a 1.1.0i build.  SHA's are shown in messages.

**Comment by lazka** [comment]
ok, here is the diff:
<details>

```diff
diff -Nur msys2/mingw64/include/openssl/bio.h other/mingw64/include/openssl/bio.h
--- msys2/mingw64/include/openssl/bio.h	2018-09-24 08:09:50.000000000 +0200
+++ other/mingw64/include/openssl/bio.h	2018-09-11 18:25:11.000000000 +0200
@@ -24,13 +24,6 @@
 #  include <openssl/e_os2.h>
 # endif
 
-/*
- * In Windows, socket-based BIO types are the SOCKET data type instead of int.
- */
-# if defined(OPENSSL_SYS_WINDOWS)
-#   include <Winsock2.h>
-# endif
-
 #ifdef  __cplusplus
 extern "C" {
 #endif
@@ -684,34 +677,20 @@
 int BIO_lookup_ex(const char *host, const char *service,
                   int lookup_type, int family, int socktype, int protocol,
                   BIO_ADDRINFO **res);
-#if defined(OPENSSL_SYS_WINDOWS)
-int BIO_sock_error(SOCKET sock);
-int BIO_socket_ioctl(SOCKET fd, long type, void *arg);
-int BIO_socket_nbio(SOCKET fd, int mode);
-int BIO_set_tcp_ndelay(SOCKET sock, int turn_on);
-#else
 int BIO_sock_error(int sock);
 int BIO_socket_ioctl(int fd, long type, void *arg);
 int BIO_socket_nbio(int fd, int mode);
-int BIO_set_tcp_ndelay(int sock, int turn_on);
-#endif /* OPENSSL_SYS_WINDOWS */
 int BIO_sock_init(void);
 # if OPENSSL_API_COMPAT < 0x10100000L
 #  define BIO_sock_cleanup() while(0) continue
 # endif
-
+int BIO_set_tcp_ndelay(int sock, int turn_on);
 
 DEPRECATEDIN_1_1_0(struct hostent *BIO_gethostbyname(const char *name))
 DEPRECATEDIN_1_1_0(int BIO_get_port(const char *str, unsigned short *port_ptr))
 DEPRECATEDIN_1_1_0(int BIO_get_host_ip(const char *str, unsigned char *ip))
-#if defined(OPENSSL_SYS_WINDOWS)
-DEPRECATEDIN_1_1_0(SOCKET BIO_get_accept_socket(char *host_port, int mode))
-DEPRECATEDIN_1_1_0(SOCKET BIO_accept(SOCKET sock, char **ip_port))
-#else
 DEPRECATEDIN_1_1_0(int BIO_get_accept_socket(char *host_port, int mode))
 DEPRECATEDIN_1_1_0(int BIO_accept(int sock, char **ip_port))
-#endif  /* OPENSSL_SYS_WINDOWS */
-
 
 union BIO_sock_info_u {
     BIO_ADDR *addr;
@@ -719,13 +698,8 @@
 enum BIO_sock_info_type {
     BIO_SOCK_INFO_ADDRESS
 };
-#if defined(OPENSSL_SYS_WINDOWS)
-int BIO_sock_info(SOCKET sock,
-                  enum BIO_sock_info_type type, union BIO_sock_info_u *info);
-#else
 int BIO_sock_info(int sock,
                   enum BIO_sock_info_type type, union BIO_sock_info_u *info);
-#endif /* OPENSSL_SYS_WINDOWS */
 
 #  define BIO_SOCK_REUSEADDR    0x01
 #  define BIO_SOCK_V6_ONLY      0x02
@@ -733,16 +707,6 @@
 #  define BIO_SOCK_NONBLOCK     0x08
 #  define BIO_SOCK_NODELAY      0x10
 
-#if defined(OPENSSL_SYS_WINDOWS)
-SOCKET BIO_socket(int domain, int socktype, int protocol, int options);
-int BIO_connect(SOCKET sock, const BIO_ADDR *addr, int options);
-int BIO_bind(SOCKET sock, const BIO_ADDR *addr, int options);
-int BIO_listen(SOCKET sock, const BIO_ADDR *addr, int options);
-SOCKET BIO_accept_ex(SOCKET accept_sock, BIO_ADDR *addr, int options);
-int BIO_closesocket(SOCKET sock);
-
-BIO *BIO_new_socket(SOCKET sock, int close_flag);
-#else
 int BIO_socket(int domain, int socktype, int protocol, int options);
 int BIO_connect(int sock, const BIO_ADDR *addr, int options);
 int BIO_bind(int sock, const BIO_ADDR *addr, int options);
@@ -751,8 +715,6 @@
 int BIO_closesocket(int sock);
 
 BIO *BIO_new_socket(int sock, int close_flag);
-#endif /* OPENSSL_SYS_WINDOWS */
-
 BIO *BIO_new_connect(const char *host_port);
 BIO *BIO_new_accept(const char *host_port);
 # endif /* OPENSSL_NO_SOCK*/
diff -Nur msys2/mingw64/include/openssl/crypto.h other/mingw64/include/openssl/crypto.h
--- msys2/mingw64/include/openssl/crypto.h	2018-09-24 08:09:50.000000000 +0200
+++ other/mingw64/include/openssl/crypto.h	2018-09-11 18:25:11.000000000 +0200
@@ -72,11 +72,7 @@
 int CRYPTO_THREAD_unlock(CRYPTO_RWLOCK *lock);
 void CRYPTO_THREAD_lock_free(CRYPTO_RWLOCK *lock);
 
-#ifdef _WIN64
-int CRYPTO_atomic_add(long long *val, int amount, int *ret, CRYPTO_RWLOCK *lock);
-#else
 int CRYPTO_atomic_add(int *val, int amount, int *ret, CRYPTO_RWLOCK *lock);
-#endif
 
 /*
  * The following can be used to detect memory leaks in the library. If
diff -Nur msys2/mingw64/include/openssl/opensslconf.h other/mingw64/include/openssl/opensslconf.h
--- msys2/mingw64/include/openssl/opensslconf.h	2018-09-24 08:09:50.000000000 +0200
+++ other/mingw64/include/openssl/opensslconf.h	2018-09-11 18:25:11.000000000 +0200
@@ -1,6 +1,6 @@
 /*
  * WARNING: do not edit!
- * Generated by Makefile from ../openssl-1.1.1/include/openssl/opensslconf.h.in
+ * Generated by Makefile from include/openssl/opensslconf.h.in
  *
  * Copyright 2016-2018 The OpenSSL Project Authors. All Rights Reserved.
  *
@@ -27,9 +27,13 @@
 #ifndef OPENSSL_SYS_MINGW64
 # define OPENSSL_SYS_MINGW64 1
 #endif
+#define OPENSSL_MIN_API 0x10000000L
 #ifndef OPENSSL_NO_MD2
 # define OPENSSL_NO_MD2
 #endif
+#ifndef OPENSSL_NO_RC5
+# define OPENSSL_NO_RC5
+#endif
 #ifndef OPENSSL_THREADS
 # define OPENSSL_THREADS
 #endif

```
</details>

**Comment by lazka** [comment]
also it's breaking ABI all over the place.. omg

**Comment by MSP-Greg** [comment]
> also it's breaking ABI

'It' being 1.1.1-1 or 1.1.1-3?  Regardless, I wondered if that would be an issue.

If something is deprecated as of 1.0.0, (since gone thru 1.0.1, 1.0.2, 1.1.0, and 1.1.1), and someone is still using it...

Ruby has been updated, so it doesn't have any issues.  Where did you see ABI issues?

**Comment by lazka** [comment]
> Where did you see ABI issues?

the patch is changing out ints for other types in public interfaces

and it includes Winsock2.h which conflicts with openssl macros

**Comment by lazka** [comment]
yeah, the winsock include is the cause of the recent build problems.

**Comment by lazka** [comment]
Fix: #4475

**Comment by MSP-Greg** [comment]
@lazka

The code I use is located at [MSP-Greg/ruby-makepkg-mingw](https://github.com/MSP-Greg/ruby-makepkg-mingw/tree/master/mingw-w64-openssl-1.1.1).

I can add output to show the config and also enable rc5.  I believe the rest should be left as is.
